subprojects/core/src/main/java/org/gradle/api/internal/FeaturePreviews.java

```diff
@@ -15,20 +15,78 @@
  */
 package org.gradle.api.internal;
 
-import org.gradle.StartParameter;
+import java.util.Collections;
+import java.util.EnumSet;
+import java.util.Set;
 
 public class FeaturePreviews {
-    private StartParameter startParameter;
 
-    public FeaturePreviews(StartParameter startParameter) {
-        this.startParameter = startParameter;
+    /**
+     * Feature previews that can be turned on.
+     * A feature that is no longer relevant will have the {@code active} flag set to {@code false}.
+     */
+    public enum Feature {
+        IMPROVED_POM_SUPPORT(true),
+        GRADLE_METADATA(true);
+
+        private final boolean active;
+
+        Feature(boolean active) {
+            this.active = active;
+        }
+
+        public boolean isActive() {
+            return active;
+        }
+    }
+
+    private final Set<Feature> activeFeatures;
+    private final EnumSet<Feature> enabledFeatures = EnumSet.noneOf(Feature.class);
+
+    public FeaturePreviews() {
+        EnumSet<Feature> tmpActiveSet = EnumSet.noneOf(Feature.class);
+        for (Feature feature : Feature.values()) {
+            if (feature.isActive()) {
+                tmpActiveSet.add(feature);
+            }
+        }
+        activeFeatures = Collections.unmodifiableSet(tmpActiveSet);
+    }
+
+    public void enableFeature(Feature feature) {
+        if (feature.isActive()) {
+            enabledFeatures.add(feature);
+        }
+    }
+
+    public boolean isFeatureEnabled(Feature feature) {
+        return feature.isActive() && enabledFeatures.contains(feature);
+    }
+
+    public void enableFeature(String name) {
+        try {
+            enableFeature(Feature.valueOf(name));
+        } catch (IllegalArgumentException e) {
+            // Re-wording to exception message to get rid of the fqcn it contains
+            throw new IllegalArgumentException("There is no feature named " + name);
+        }
     }
 
-    public boolean isAdvancedPomSupportEnabled() {
-        return startParameter.isAdvancedPomSupport();
+    public boolean isFeatureEnabled(String name) {
+        try {
+            return isFeatureEnabled(Feature.valueOf(name));
+        } catch (IllegalArgumentException e) {
+            // Re-wording to exception message to get rid of the fqcn it contains
+            throw new IllegalArgumentException("There is no feature named " + name);
+        }
```

Comment:
Nit: I'd extract a method `Feature featureNamed(String name)` (or a static method `Feature.withName()`) to get rid of the duplicated catch here.